subsys/net/lib/sockets/socketpair.c

```diff
@@ -0,0 +1,1147 @@
+/*
+ * Copyright (c) 2020 Friedt Professional Engineering Services, Inc
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <fcntl.h>
+
+/* Zephyr headers */
+#include <logging/log.h>
+LOG_MODULE_REGISTER(net_spair, CONFIG_NET_SOCKETS_LOG_LEVEL);
+
+#undef LOG_DBG
+#define LOG_DBG(fmt, args...) printk("%s(): %d: " fmt "\n", __func__, __LINE__, ##args)
+
+#undef LOG_ERR
+#define LOG_ERR LOG_DBG
+
+#include <kernel.h>
+#include <net/socket.h>
+#include <syscall_handler.h>
+#include <sys/__assert.h>
+#include <sys/fdtable.h>
+
+#include "sockets_internal.h"
+
+enum {
+	SPAIR_SIG_CANCEL,
+	SPAIR_SIG_DATA,
+};
+
+enum {
+	SPAIR_FLAG_NONBLOCK = (1 << 0),
+};
+
+#define SPAIR_FLAGS_DEFAULT 0
+
+struct spair {
+	int remote;
+	u32_t flags;
+	struct k_sem sem;
+	struct k_pipe recv_q;
+	struct k_poll_signal write_signal;
+	struct k_poll_signal read_signal;
+	u8_t buf[CONFIG_NET_SOCKETPAIR_BUFFER_SIZE];
+};
+
+/* forward declaration */
+static const struct socket_op_vtable spair_fd_op_vtable;
+
+#undef sock_is_nonblock
+/**
+ * Determine if a @ref spair is in non-blocking mode
+ */
+static inline bool sock_is_nonblock(const struct spair *spair)
+{
+	return !!(spair->flags & SPAIR_FLAG_NONBLOCK);
+}
+
+/**
+ * Determine if a @ref spair is connected
+ */
+static inline bool sock_is_connected(const struct spair *spair)
+{
+	const struct spair *remote = z_get_fd_obj(spair->remote,
+		(const struct fd_op_vtable *)&spair_fd_op_vtable, 0);
+
+	if (!remote) {
+		return false;
+	}
+
+	return true;
+}
+
+#undef sock_is_eof
+/**
+ * Determine if a @ref spair has encountered end-of-file
+ */
+static inline bool sock_is_eof(const struct spair *spair)
+{
+	return !sock_is_connected(spair);
+}
+
+/**
+ * Determine bytes available to write
+ *
+ * Specifically, this function calculates the number of bytes that may be
+ * written to a given @ref spair without blocking
+ */
+static inline size_t spair_write_avail(struct spair *spair)
+{
+	struct spair *const remote = z_get_fd_obj(spair->remote,
+		(const struct fd_op_vtable *)&spair_fd_op_vtable, 0);
+
+	struct k_pipe *const p = (remote == NULL)
+		? NULL
+		: &remote->recv_q;
+
+	if (remote == NULL) {
+		return 0;
+	}
+
+	LOG_DBG("pipe: %p N: %u u: %u r: %u w: %u", p, p->size,
```

Comment:
This logging apparently should be removed from final version.

Reply:
Do `LOG_DBG()` messages not get compiled-out when it's not selected in Kconfig? It should be helpful for debugging, which is why I left it in there.

Reply:
> Do LOG_DBG() messages not get compiled-out when it's not selected in Kconfig? It should be helpful for debugging, which is why I left it in there.

As you remember, the original concern with the code which was there was encapsulation violation - user of k_pipe should use functions defined on it, not access its internal structure directly. You fixed that, yet still access internal fields in logging code. I personally can easily buy "debugging" argument, but it's still encapsulation violation in the end.

Reply:
OK